code/modules/reagents/chemistry/recipes.dm

```diff
@@ -57,6 +90,99 @@
+ * * react_volume - volume created across the whole reaction
+ */
+/datum/chemical_reaction/proc/reaction_finish(datum/reagents/holder, react_vol)
+	//failed_chem handler
+	for(var/id in results)
+		var/datum/reagent/R = holder.has_reagent(id)
+		if(!R)
+			continue
+		if(R.purity < purity_min)
+			var/cached_volume = R.volume
+			holder.remove_reagent(R.type, cached_volume, FALSE)
+			holder.add_reagent(R.failed_chem, cached_volume, FALSE, added_purity = 1)
+			SSblackbox.record_feedback("tally", "chemical_reaction", 1, "[type] failed reactions")
+			continue
+
+		//REACTION_CLEAR handler
+		if(reaction_flags & (REACTION_CLEAR_IMPURE | REACTION_CLEAR_INVERSE))
+			if(R.purity == 1) //Failures can delete R
+				continue
+
+			var/cached_volume = R.volume
+			if((reaction_flags & REACTION_CLEAR_INVERSE) && R.inverse_chem)
+				if(R.inverse_chem_val > R.purity)
+					holder.remove_reagent(R.type, cached_volume, FALSE)
+					holder.add_reagent(R.inverse_chem, cached_volume, FALSE, added_purity = 1)
+
+			if((reaction_flags & REACTION_CLEAR_IMPURE) && R.impure_chem)
+				var/impureVol = cached_volume * (1 - R.purity)
+				holder.remove_reagent(R.type, (impureVol), FALSE)
+				holder.add_reagent(R.impure_chem, impureVol, FALSE, added_purity = 1)
+				R.creation_purity = R.purity
+				R.purity = 1
+
+
+
+/**
+ * Occurs when a reation is overheated (i.e. past it's overheatTemp)
+ * Will be called every tick in the reaction that it is overheated
+ * If you want this to be a once only proc (i.e. the reaction is stopped after) set reaction.toDelete = TRUE
+ * The above is useful if you're writing an explosion
+ * By default the parent proc will reduce the final yield slightly. If you don't want that don't add ..()
+ *
+ * Arguments:
+ * * holder - the datum that holds this reagent, be it a beaker or anything else
+ * * equilibrium - the equilibrium datum that contains the equilibrium reaction properties and methods
+ */
+/datum/chemical_reaction/proc/overheated(datum/reagents/holder, datum/equilibrium/equilibrium)
+	for(var/id in results)
+		var/datum/reagent/R = holder.get_reagent(id)
+		if(!R)
+			return
+		R.volume =  round((R.volume*0.98), 0.01) //Slowly lower yield per tick
+
+/**
+ * Occurs when a reation is too impure (i.e. it's below purity_min)
+ * Will be called every tick in the reaction that it is too impure
+ * If you want this to be a once only proc (i.e. the reaction is stopped after) set reaction.toDelete = TRUE
+ * The above is useful if you're writing an explosion
+ * By default the parent proc will reduce the purity of all reagents in the beaker slightly. If you don't want that don't add ..()
+ *
+ * Arguments:
+ * * holder - the datum that holds this reagent, be it a beaker or anything else
+ * * equilibrium - the equilibrium datum that contains the equilibrium reaction properties and methods
+ */
+/datum/chemical_reaction/proc/overly_impure(datum/reagents/holder, datum/equilibrium/equilibrium)
```

Comment:
Why is the equilibrium datum being passed here as an arg?

How knowledgeable are you on components? This datum is a very good candidate for it.

Reply:
Unfortunately I know nothing about byond components and they're a little mystifying when I have to use them.

Reply:
If you want, poke me on discord and I can guide you through them. They might look cryptic at first, but they're pretty simple at the end of the day, and it shouldn't take too long to learn how to use them.

If you have enough on your plate already, however, as this is a big PR and it can be taxing, feel free to resolve this.

A component would be good but it's not strictly necessary.